controllers/vpcresources/cninode_controller.go

```diff
@@ -0,0 +1,128 @@
+	ec2API "github.com/aws/amazon-vpc-resource-controller-k8s/pkg/aws/ec2/api"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/aws/ec2/api/cleanup"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/config"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/finalizer"
+)
+
+// CNINodeReconciler reconciles a CNINode object
+type CNINodeReconciler struct {
+	client.Client
+	Scheme      *runtime.Scheme
+	Context     context.Context
+	Log         logr.Logger
+	EC2Wrapper  ec2API.EC2Wrapper
+	ClusterName string
+	finalizers  finalizer.Finalizers
+}
+
+// nodeTerminationFinalizer will run the finalizer routine to clean up leaked resources on the node if CNINode is being deleted
+// CNINode is deleted when the parent node object is delete ie when the EC2 instance is terminated
+type nodeTerminationFinalizer struct {
+	EC2Wrapper ec2API.EC2Wrapper
+}
+
+var _ finalizer.Finalizer = (*nodeTerminationFinalizer)(nil)
+
+//+kubebuilder:rbac:groups=vpcresources.k8s.aws,resources=cninodes,verbs=get;list;watch;create;update;patch;
+
+// Reconcile handles CNINode create/update/delete events
+// Reconciler will add the finalizer and cluster name tag if it does not exist and finalize on CNINode on deletion to clean up leaked resource on node
+func (r *CNINodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	cniNode := &v1alpha1.CNINode{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cniNode); err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("CNINode is deleted, will be re-created if the node exists", "CNINode", req.NamespacedName)
+		}
+		// Ignore not found error; CNINode will be re-created on node reconcile if the node is not terminated
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Add cluster name tag if it does not exist
+	clusterNameTagKey := fmt.Sprintf(config.ClusterNameTagKeyFormat, r.ClusterName)
+	val, ok := cniNode.Spec.Tags[clusterNameTagKey]
+	if !ok || val != config.ClusterNameTagValue {
+		newCNINode := cniNode.DeepCopy()
+		if len(newCNINode.Spec.Tags) != 0 {
+			newCNINode.Spec.Tags[clusterNameTagKey] = config.ClusterNameTagValue
+		} else {
+			newCNINode.Spec.Tags = map[string]string{
+				clusterNameTagKey: config.ClusterNameTagValue,
+			}
+		}
+
+		if err := r.Client.Patch(ctx, newCNINode, client.MergeFrom(cniNode)); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to patch CNINode %v", err)
+		}
+	}
+
+	// Finalize will add the finalizer if it does not exist and object is not being delete, and will remove the finalizer if it exists
+	// and object is being deleted. Also sets result.Updated to true so we need to update the object
+	finalizationRes, err := r.finalizers.Finalize(ctx, cniNode)
```

Comment:
I think we should add deletion protection for CNINode here. If the finalizer is updated (deletion timestamp is added), we can copy and recreate the CNINode iif the node is still existing (which means accidently or incorrectly deleted the CNINode object). Then we should have all data from the original object and create a copy soon. Something like
1, deletion timestamp is found
2, check if the node is deleted
3, the node is gone --> continue finalizing the CNINode
4, the node is still existing --> giving up finalizing and copying and recreating the CNINode

We have to implementation the protection until upstream [KEP for in-use protection](https://github.com/kubernetes/enhancements/pull/2840) is implemented and graduated.